File: src/com/github/luomingxu/idea/action/BaseAction.java
package com.github.luomingxu.idea.action;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleUtil;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;

public class BaseAction extends AnAction {

    public void action(Project project, Module module, PsiFile psiFile, PsiElement psiElement) {}

    @Override
    public void actionPerformed(AnActionEvent e) {
        DataContext dataContext = e.getDataContext();
        Project project = CommonDataKeys.PROJECT.getData(dataContext);
        assert project != null;
        PsiFile psiFile = CommonDataKeys.PSI_FILE.getData(dataContext);
        assert psiFile != null;
        Module module = ModuleUtil.findModuleForFile(psiFile);
        assert module != null;
        PsiElement psiElement = CommonDataKeys.PSI_ELEMENT.getData(dataContext);

        action(project, module, psiFile, psiElement);
    }
}

File: src/com/github/luomingxu/idea/action/BaseAction.kt
package com.github.luomingxu.idea.action

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.module.Module
import com.intellij.openapi.module.ModuleUtil
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile

open class BaseAction : AnAction() {
    open fun action(project: Project?, module: Module?, psiFile: PsiFile?, psiElement: PsiElement?) {}

    override fun actionPerformed(e: AnActionEvent) {
        val dataContext = e.dataContext
        val project = CommonDataKeys.PROJECT.getData(dataContext)
        val psiFile = CommonDataKeys.PSI_FILE.getData(dataContext)
        val module = psiFile?.let { ModuleUtil.findModuleForFile(it) }
        val psiElement = CommonDataKeys.PSI_ELEMENT.getData(dataContext)
        action(project, module, psiFile, psiElement)
    }

    fun <T1, T2> ifNotNull(value1: T1?, value2: T2?, bothNotNull: (T1, T2) -> (Unit)) {
        if (value1 != null && value2 != null) {
            bothNotNull(value1, value2)
        }
    }
}
